ksqldb-streams/src/main/java/io/confluent/ksql/execution/streams/materialization/KsqlMaterialization.java

```diff
@@ -128,148 +131,157 @@ public MaterializedWindowedTable windowed() {
+    KsqlMaterializedTable(final StreamsMaterializedTable table) {
       this.table = requireNonNull(table, "table'");
     }
 
     @Override
-    public KsMaterializedQueryResult<Row> get(final GenericKey key, final int partition) {
-      if (transforms.isEmpty()) {
-        return table.get(key, partition);
-      }
+    public KsqlMaterializedQueryResult<Row> get(
+        final GenericKey key,
+        final int partition,
+        final Optional<ConsistencyOffsetVector> consistencyVector
+    ) {
+      final Optional<Position> position = consistencyVector
+          .map(offsetVector -> Position.fromMap(offsetVector.getOffsetVector()));
+      final KsMaterializedQueryResult<Row> result = table.get(key, partition, position);
+      updateConsistencyVector(consistencyVector, result.getPosition());
 
-      final KsMaterializedQueryResult<Row> result = table.get(key, partition);
-      final Iterator<Row> iterator = Streams.stream(result.getRowIterator())
-          .map(row -> filterAndTransform(row.key(), getIntermediateRow(row), row.rowTime())
-              .map(v -> row.withValue(v, schema())))
-          .filter(Optional::isPresent)
-          .map(Optional::get)
-          .iterator();
-
-      if (result.getPosition().isPresent()) {
-        return KsMaterializedQueryResult.rowIteratorWithPosition(
-            iterator, result.getPosition().get());
+      if (transforms.isEmpty()) {
+        return KsqlMaterializedQueryResult.rowIterator(result.getRowIterator());
       } else {
-        return KsMaterializedQueryResult.rowIterator(iterator);
+        final Iterator<Row> iterator = Streams.stream(result.getRowIterator())
+            .map(row -> filterAndTransform(row.key(), getIntermediateRow(row), row.rowTime())
+                .map(v -> row.withValue(v, schema())))
+            .filter(Optional::isPresent)
+            .map(Optional::get)
+            .iterator();
+        return KsqlMaterializedQueryResult.rowIterator(iterator);
       }
     }
 
     @Override
-    public KsMaterializedQueryResult<Row> get(final int partition) {
-      if (transforms.isEmpty()) {
-        return table.get(partition);
-      }
-
-      final KsMaterializedQueryResult<Row> result = table.get(partition);
-
-      final Iterator<Row> iterator = Streams.stream(result.getRowIterator())
-          .map(row -> filterAndTransform(row.key(), getIntermediateRow(row), row.rowTime())
-              .map(v -> row.withValue(v, schema())))
-          .filter(Optional::isPresent)
-          .map(Optional::get)
-          .iterator();
+    public KsqlMaterializedQueryResult<Row> get(
+        final int partition,
+        final Optional<ConsistencyOffsetVector> consistencyVector
+    ) {
+      final Optional<Position> position = consistencyVector
+          .map(offsetVector -> Position.fromMap(offsetVector.getOffsetVector()));
+      final KsMaterializedQueryResult<Row> result = table.get(partition, position);
```

Comment:
to make sure i'm reviewing the pr correctly, this (and the same part below) is the "meat" of it, right? passing in `position` to the underlying table?

(can be done in a different PR) do we need ot have ConsistencyVector now that we have Position?

Reply:
We decided to keep both for now among other reasons so that we can make changes to the ConsistencyVector faster than what would be possible with streams.